### PR TITLE
Refactor TransportSocketFactoryContext and Cluster interfaces.

### PR DESCRIPTION
**Pull request by JimmyCYJ**
Refactor TransportSocketFactoryContext and Cluster interfaces.

Description:
According to #3700 , we introduce TransportSocketFactoryContextImpl that simplifies the interfaces of 
ListenerImpl,  ClusterImplBase and its derived classes. This PR implements TransportSocketFactoryContextImpl, and refactors ListenerImpl and Cluster interfaces. This PR is split out of #3700.

Risk Level: Low
Signed-off-by: Jimmy Chen jimmychen.0102@gmail.com

### CODE REVIEW

**Comment by JimmyCYJ** [comment]
@lizan @qiwzhang please take a look. Thanks!

File: include/envoy/server/transport_socket_config.h [hunk 1]
```diff
@@ -33,6 +38,31 @@ class TransportSocketFactoryContext {
...
+
+  /**
+   * @return the instance of ClusterManager.
+   */
+  virtual Upstream::ClusterManager& clusterManager() PURE;
+
+  /**
+   * @return information about the local environment the server is running in.
+   */
+  virtual const LocalInfo::LocalInfo& local_info() PURE;
```
Comment (lizan) [inline include/envoy/server/transport_socket_config.h]: `localInfo`

File: source/common/upstream/eds.h [hunk 2]
```diff
@@ -19,10 +19,9 @@ class EdsClusterImpl : public BaseDynamicClusterImpl,
                        Config::SubscriptionCallbacks<envoy::api::v2::ClusterLoadAssignment> {
 public:
   EdsClusterImpl(const envoy::api::v2::Cluster& cluster, Runtime::Loader& runtime,
-                 Stats::Store& stats, Ssl::ContextManager& ssl_context_manager,
-                 const LocalInfo::LocalInfo& local_info, ClusterManager& cm,
-                 Event::Dispatcher& dispatcher, Runtime::RandomGenerator& random,
-                 bool added_via_api);
+                 bool added_via_api,
```
Comment (lizan) [inline source/common/upstream/eds.h]: Can we keep `added_via_api` as the last parameter?

File: source/common/upstream/eds.h [hunk 3]
```diff
@@ -19,10 +19,9 @@ class EdsClusterImpl : public BaseDynamicClusterImpl,
...
+                 Server::Configuration::TransportSocketFactoryContext& factory_context,
+                 Stats::ScopePtr stats_scope);
```
Comment (lizan) [inline source/common/upstream/eds.h]: `ScopePtr&&`? (also for other ClusterImpl)

File: source/common/upstream/upstream_impl.h [hunk 4]
```diff
@@ -406,26 +400,24 @@ class ClusterInfoImpl : public ClusterInfo,
   const std::chrono::milliseconds connect_timeout_;
   absl::optional<std::chrono::milliseconds> idle_timeout_;
   const uint32_t per_connection_buffer_limit_bytes_;
-  Stats::ScopePtr stats_scope_;
-  mutable ClusterStats stats_;
   Stats::IsolatedStoreImpl load_report_stats_store_;
   mutable ClusterLoadReportStats load_report_stats_;
   Network::TransportSocketFactoryPtr transport_socket_factory_;
+  Stats::ScopePtr stats_scope_;
```
Comment (lizan) [inline source/common/upstream/upstream_impl.h]: why did you change the order?

**Comment by JimmyCYJ** [comment]
@lizan Please take a look. Thanks!

File: test/common/upstream/upstream_impl_test.cc [hunk 5]
```diff
@@ -1150,16 +1280,27 @@ TEST(StaticClusterImplTest, UnsupportedLBType) {
   }
   )EOF";
 
-  EXPECT_THROW(StaticClusterImpl(parseClusterFromJson(json), runtime, stats, ssl_context_manager,
-                                 local_info, cm, false),
-               EnvoyException);
+  EXPECT_THROW(
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply test/common/upstream/upstream_impl_test.cc]: Done. Please take a look.
